### PR TITLE
httpd: Fix a few things in git router

### PR DESCRIPTION
**Pull request by xphoniex**

### CODE REVIEW

File: radicle-httpd/src/lib.rs [hunk 1]
```diff
@@ -95,8 +88,60 @@ pub async fn run(options: Options) -> anyhow::Result<()> {
...
+    #[tokio::test]
+    async fn test_invalid_route_returns_404() {
+        let tmp = tempfile::tempdir().unwrap();
+        let _ctx = test::seed(tmp.path());
+        let app = super::router(
+            super::Options {
+                aliases: HashMap::new(),
+                listen: SocketAddr::from(([0, 0, 0, 0], 8080)),
+            },
+            Home::new(tmp.path().join("radicle")).unwrap(),
```
Comment (xphoniex) [inline radicle-httpd/src/lib.rs:138]: sub-routes take a `ctx` which has profile in it, and are designed to be easily testable, and here, to be able to re-produce the same system we have to pass `home` ourselves. 

for this to work, I had to also add `load_at(home: Home) -> Result<Self, Error>` to [profile.rs](https://github.com/radicle-dev/heartwood/blob/03c636c65d757ab7008ee3c65963dd585979dc64/radicle/src/profile.rs).

**Review by xphoniex — COMMENTED** [review]
this is only for test, have two other commits which I'll add.

thoughts on this?

File: radicle-httpd/src/lib.rs [hunk 2]
```diff
@@ -95,8 +88,60 @@ pub async fn run(options: Options) -> anyhow::Result<()> {
...
+
+    let api_router = api::router(ctx);
+    let git_router = git::router(profile.clone(), options.aliases);
+    let raw_router = raw::router(profile);
+
+    let app = Router::new()
+        .merge(git_router)
+        .nest("/api", api_router)
+        .nest("/raw", raw_router);
+
```
Comment (xphoniex) [inline radicle-httpd/src/lib.rs:113]: also I tried to get it as close as possible to the production router, but it doesn't play nice as once you call
```rust
.into_make_service_with_connect_info(..)
```

you can no longer use this in test:
```rust
.layer(MockConnectInfo(SocketAddr::from(([0, 0, 0, 0], 8080))));
```

**Comment by xphoniex** [comment]
the failing test doesn't seem to have much to do with my commit :/ ?

**Comment by cloudhead** [comment]
git-daemon port collision.. just re-ran it.

File: radicle-httpd/src/error.rs [hunk 3]
```diff
@@ -24,40 +24,56 @@ pub enum Error {
...
     InvalidHeaderValue(#[from] axum::http::header::InvalidHeaderValue),
+}
+
+impl GitError {
+    pub fn status(&self) -> http::StatusCode {
+        match self {
+            GitError::ServiceUnavailable(_) => http::StatusCode::SERVICE_UNAVAILABLE,
+            GitError::Id(_) => http::StatusCode::NOT_FOUND,
+            GitError::NotFound => http::StatusCode::NOT_FOUND,
+            _ => http::StatusCode::INTERNAL_SERVER_ERROR,
```
[the remark this answers is not in the document]
Comment (cloudhead) [reply radicle-httpd/src/error.rs:42]: Yeah.

File: radicle/src/profile.rs [hunk 4]
```diff
@@ -84,6 +84,11 @@ impl Profile {
 
     pub fn load() -> Result<Self, Error> {
         let home = self::home()?;
+
+        Self::load_at(home)
+    }
+
+    pub fn load_at(home: Home) -> Result<Self, Error> {
```
Comment (cloudhead) [inline radicle/src/profile.rs:91]: Can be reverted

File: radicle-httpd/src/lib.rs [hunk 5]
```diff
@@ -95,8 +88,56 @@ pub async fn run(options: Options) -> anyhow::Result<()> {
...
-    axum::Server::bind(&options.listen)
+    axum::Server::bind(&listen)
         .serve(app)
         .await
         .map_err(anyhow::Error::from)
 }
+
+/// Create a router consisting of other sub-routers.
+fn router(options: Options, profile: Profile) -> anyhow::Result<Router> {
+    tracing::info!("using radicle home at {}", profile.home().display());
```
Comment (cloudhead) [inline radicle-httpd/src/lib.rs:99]: I think this log line should be in the `run` function.

Comment (xphoniex) [reply radicle-httpd/src/lib.rs]: yup, my thoughts too. https://github.com/radicle-dev/heartwood/pull/422/commits/75463a7872255a89490b362f7e11962c56cd15f2#diff-f5911c8765b18e1a9d9d0957001f22ff11cc9c0f5749f9293fae412ccd681362R56

File: radicle/src/profile.rs (hunk 4 above)
Comment (xphoniex) [reply radicle/src/profile.rs]: reverted in https://github.com/radicle-dev/heartwood/pull/422/commits/75463a7872255a89490b362f7e11962c56cd15f2#diff-f5911c8765b18e1a9d9d0957001f22ff11cc9c0f5749f9293fae412ccd681362R56

**Comment by xphoniex** [comment]
also added https://github.com/radicle-dev/heartwood/pull/422/commits/e0cbc9104f83768ab3fcc32f723b636d34e3705a , the only thing remaining is remote e.g. [here](https://github.com/radicle-dev/heartwood/blob/e0cbc9104f83768ab3fcc32f723b636d34e3705a/radicle-httpd/src/api/v1/projects.rs#L993)

**Review by cloudhead — APPROVED** [review]

**Pull request merged**